src/UsgsAstroLsSensorModel.cpp

```diff
@@ -654,112 +652,91 @@ csm::ImageCoord UsgsAstroLsSensorModel::groundToImage(
+   c = m_detectorLineCoeffs[3*referenceIndex + 2];
 
-   // Reintersect to ensure the image point actually views the ground point.
-   csm::EcefCoord calculatedPoint = imageToGround(calculatedPixel, height);
-   double dx = ground_pt.x - calculatedPoint.x;
-   double dy = ground_pt.y - calculatedPoint.y;
-   double dz = ground_pt.z - calculatedPoint.z;
-   double len = dx * dx + dy * dy + dz * dz;
+   double d0 = distanceToLine(p0[0], p0[1], a, b, c);
+   double di = distanceToLine(pi[0], pi[1], a, b, c);
+   double delta = 0;
+   int count = 0;
 
-   if (achieved_precision) {
-      *achieved_precision = sqrt(len);
-   }
+   while (abs(di) > m_averageDetectorSize) {
+     delta = int(abs(di) / m_averageDetectorSize);
 
-   double preSquare = desired_precision * desired_precision;
-   if (warnings && (desired_precision > 0.0) && (preSquare < len)) {
-     MESSAGE_LOG(m_logger, "groundToImage: Desired precision not achieved {}", preSquare)
-     std::stringstream msg;
-     msg << "Desired precision not achieved. ";
-     msg << len << "  " << preSquare << "\n";
-     warnings->push_back(csm::Warning(csm::Warning::PRECISION_NOT_MET,
-                         msg.str().c_str(),
-                         "UsgsAstroLsSensorModel::groundToImage()"));
+     if (d0 * di > 0) {
+       delta = delta;
+     }
+     else if (d0 * di <= 0) {
+       delta = -delta;
+     }
+
+     approxPt.line += delta;
+     timei = getImageTime(approxPt);
+     pi = computeDetectorView(timei, groundPt, adj, desiredPrecision);
+     di = distanceToLine(pi[0], pi[1], a, b, c);
+
+     count++;
+
+     if (count > 30) {
+       break;
+     }
    }
-   MESSAGE_LOG(m_logger, "groundToImage: Final pixel line {}, sample {}", calculatedPixel.line, calculatedPixel.samp)
+   timei = getImageTime(approxPt);
+   std::vector<double> pi1 = computeDetectorView(timei, groundPt, adj, desiredPrecision);
+   double di1 = distanceToLine(pi1[0], pi1[1], a, b, c);
+   double line = approxPt.line + (abs(di) / (abs(di) + abs(di1)));
+
+   // Invert distortion
+   double distortedFocalX, distortedFocalY;
+   applyDistortion(pi[0], pi[1], distortedFocalX, distortedFocalY,
+                   m_opticalDistCoeffs, m_distortionType, desiredPrecision);
+
+   // Convert to detector line and sample
+   double detectorLine = m_iTransL[0]
+                       + m_iTransL[1] * distortedFocalX
+                       + m_iTransL[2] * distortedFocalY;
+   double detectorSample = m_iTransS[0]
+                         + m_iTransS[1] * distortedFocalX
+                         + m_iTransS[2] * distortedFocalY;
 
-   return calculatedPixel;
+   // Convert to image sample line
+   line += detectorLine + m_detectorLineOrigin;
+   double sample = (detectorSample + m_detectorSampleOrigin - m_startingSample)
+                 / m_detectorSampleSumming;
+   MESSAGE_LOG(m_logger, "computeViewingPixel: image line sample {} {}",
+                                line, sample)

```

Comment:
For test-ability, I would make these two steps a separate function that we can test in isolation.